packages/hydrated_bloc/lib/src/hydrated_cubit.dart

```diff
@@ -116,25 +120,118 @@ mixin HydratedMixin<State> on Cubit<State> {
     try {
       final stateJson = storage.read(storageToken) as Map<dynamic, dynamic>;
       if (stateJson == null) return _state = super.state;
-      return _state = fromJson(Map<String, dynamic>.from(stateJson));
+      return _state = fromJson(
+        _cast<Map<String, dynamic>>(_traverseRead(stateJson)),
+      );
     } on dynamic catch (error, stackTrace) {
       onError(error, stackTrace);
       return _state = super.state;
     }
   }
 
+  dynamic _traverseRead(dynamic value) {
+    if (value is Map) {
+      final map = <String, dynamic>{};
+      value.forEach((dynamic key, dynamic value) {
+        map[_cast<String>(key)] = _traverseRead(value);
+      });
+      return map;
+    }
+    if (value is List) {
+      for (var i = 0; i < value.length; i++) {
+        value[i] = _traverseRead(value[i]);
+      }
+    }
+    return value;
+  }
+
+  T _cast<T>(dynamic x) => x is T ? x : null;
+
   @override
   void onChange(Change<State> change) {
     if (storage == null) throw const StorageNotFound();
     final state = change.nextState;
-    final stateJson = toJson(state);
+    final stateJson = _toJson(state);
```

Comment:
should it be also here?

Reply:
yeah I think this should also be wrapped in a try/catch